### ISSUE TITLE
[Meteor 1.4.1.1] Mongo Upsert with multi-attribute query results in undefined object in mongo_driver.js:636

### ISSUE THREAD

**Issue by BigDSK** [meteor/meteor#7758]
Environment:
- Meteor 1.4.1.1
- OS: MacOSX 10.11.6
- Dependencies: aldeed:simple-schema; aldeed:collections2

Upsert code that worked in versions 1.3.X and below is now failing in 1.4.1.1 with an error in mongo_driver.js:636. The error and stack are below. An example statement is here (see repo below for full reproduction):

```
  Sample.upsert({
      network_id: new Mongo.ObjectID('44915733af80844fa1cef07a'),
      collection_name: 'blah',
      doc_id: new Mongo.ObjectID('44915733af80844fa1cef07b')
    },
    {
      $set: { thing_to_update: 'item1' }
    });
```

Full reproducible repo is here:

[https://github.com/BigDSK/meteor-upsert-error](https://github.com/BigDSK/meteor-upsert-error)

Please view [/server/main.js](https://github.com/BigDSK/meteor-upsert-error/blob/master/server/main.js#L42) for the collection definition and offending code. Workaround is to replace upsert with findOne()/insert()/update() series.

```
Exception in Mongo write: TypeError: Cannot read property 'indexOf' of undefined
     at Function.MongoConnection._isCannotChangeIdError (packages/mongo/mongo_driver.js:636:7)
     at packages/mongo/mongo_driver.js:751:47
     at runWithEnvironment (packages/meteor/dynamics_nodejs.js:110:1)
     at packages/meteor/dynamics_nodejs.js:123:1
     at handleCallback (/Users/dennis/.meteor/packages/npm-mongo/.1.5.49.12xux8a++os+web.browser+web.cordova/npm/node_modules/mongodb/lib/utils.js:96:12)
     at /Users/dennis/.meteor/packages/npm-mongo/.1.5.49.12xux8a++os+web.browser+web.cordova/npm/node_modules/mongodb/lib/collection.js:1043:20
     at executeWrite (/Users/dennis/.meteor/packages/npm-mongo/.1.5.49.12xux8a++os+web.browser+web.cordova/npm/node_modules/mongodb-core/lib/wireprotocol/3_2_support.js:80:5)
     at [object Object].WireProtocol.update (/Users/dennis/.meteor/packages/npm-mongo/.1.5.49.12xux8a++os+web.browser+web.cordova/npm/node_modules/mongodb-core/lib/wireprotocol/3_2_support.js:93:3)
     at [object Object].Server.update (/Users/dennis/.meteor/packages/npm-mongo/.1.5.49.12xux8a++os+web.browser+web.cordova/npm/node_modules/mongodb-core/lib/topologies/server.js:559:35)
     at [object Object].Server.update (/Users/dennis/.meteor/packages/npm-mongo/.1.5.49.12xux8a++os+web.browser+web.cordova/npm/node_modules/mongodb/lib/server.js:386:17)
```

**Comment by abernix** [comment]
I apologize its taken so long to look at this.  Thank you for the very clear reproduction on this error! 🙏

This does appear to be a bug, and possibly two bugs with the Mongo driver.

By running from checkout and adding a `debugger` statement [here](https://github.com/meteor/meteor/blob/e86ef3cb47fb4f44a06b35433d94cae4602e7bcc/packages/mongo/mongo_driver.js#L749) the error can be seen when the `upsert` is executed.  It's passing through [this block](https://github.com/meteor/meteor/blob/e86ef3cb47fb4f44a06b35433d94cae4602e7bcc/packages/mongo/mongo_driver.js#L749-L765) where the `err` is set to a JavaScript `Error` that looks like this:

![image](https://cloud.githubusercontent.com/assets/841294/19290197/39b22cda-9017-11e6-9969-f299b15ee584.png)

At first glance, I question:

 * Why is the object/BSON so wrong?  Definitely looks like it's being converted incorrectly somewhere.
 * Why `err.err` was changed to `err.errmsg` [in this commit](https://github.com/meteor/meteor/commit/72fb8690c0cb69d0453a7df91979c7d0d80020d1#diff-a9d0b23557a439d95f0ceea868fb961bL604) for `_isCannotChangeIdError` during the Mongo upgrade.
 * Why `err.message` wasn't used instead, which would have properly caught the true error (i.e. `Invalid ObjectID`).

The [Native Node Mongo driver API](https://mongodb.github.io/node-mongodb-native/2.2/api/MongoError.html) says that this should be `message`, but the [Mongo `getLastError` docs](https://docs.mongodb.com/manual/reference/command/getLastError/#getLastError.err) show returning the `err` and `errmsg`.  Perhaps this was the point of confusion?

I'm not familiar with the path taken to cause this error or if it would be expected for this to be a Mongo DB error or a Mongo Driver error.

These are just my notes, but this will need further investigation.

@Fabs Any thoughts on this?

**Comment by Fabs** [comment]
I will definitely take a look at it tonight, and see if I can contribute with something.

**Comment by Fabs** [comment]
Sorry. I got to it just today :).

Regarding the `errmsg`. I believe it is a fix I applied later for [accounts-base.](https://github.com/meteor/meteor/search?utf8=%E2%9C%93&q=errmsg&type=Code). It made sense to me, to apply at this place as well. If I remember right, changing the driver broke the indexOf on the err.err. The reason I did not use `err.mssage` is unclear to me. Probably a mistake. 

For the ID, I think that the format of the Id also changed. And I had failing tests for a long time. My guess here it is because the id the driver returns changed format as well. When you use a Collection, then the api is right, because we return *knownId* from the mod._id or selector._id (query basic) or no ID at all. But internally when the upsert will use update or insert, it will get a representation of the ID it does not know how to use. Then maybe `replaceTypes` do something wrong with it, or something else bad happens.

**Comment by Fabs** [comment]
If nobody comes to it during this week. I will make free time on the weekend to write a (some) tests that fails based on the reproduction @BigDSK gave to us, and try to fix it.

**Comment by jmfly** [comment]
Any thoughts on this issue?

**Comment by Fabs** [comment]
I gave it a shot and it was not straightforward to fix. I will post here if I can take of a weekend to solve it.

**Comment by tekstop** [comment]
Started running into this when we updated our app to 1.4.1

**Comment by bpartridge** [comment]
Seems this is fixed in 1.5.2.2 - can someone else confirm? Very possible that https://github.com/meteor/meteor/commit/8a7db6bdc55a07d53816dd2a4c8c0572c50dec25#diff-a9d0b23557a439d95f0ceea868fb961b inadvertently fixed this. I don't know enough about the internals to know if that's plausible that that change would have fixed that gnarly BSON conversion, so I'd hesitate suggesting that we close this without another look.

**Comment by bpartridge** [comment]
For the record, ran into this in production running 1.4.3.1 - it (was?) definitely real. We haven't made the leap to 1.5 yet, but this might make it necessary.

**Comment by hwillson** [comment]
I can confirm that the original reproduction included with this issue is working with 1.5.2.2. Thanks for pointing this out @bpartridge!
